### PR TITLE
thread safety

### PR DESCRIPTION
**Pull request by osheroff**
old code, not so thread-safe.

### CODE REVIEW

**Comment by osheroff** [comment]
ping @grosser

File: lib/active_record/comments.rb [hunk 1]
```diff
@@ -68,7 +68,7 @@ def with_comment_sql(sql)
       private
 
       def current_comment
-        @comment.join(" ") if @comment.present?
+        Thread.current[:ar_comments].join(" ") if Thread.current[:ar_comments].present?
```
Comment (grosser) [inline lib/active_record/comments.rb]: might be dryer by using something like this:
```Ruby
def comments
  Thread.current[:ar_comments] ||= []
end
```

**Comment by grosser** [comment]
can I haz tests ?

**Comment by osheroff** [comment]
updated

File: lib/active_record/comments.rb [hunk 2]
```diff
@@ -53,11 +53,11 @@ def install(adapter)
 
     class << self
       def comment(comment)
-        @comment ||= []
-        @comment << comment
+        current_comments
```
Comment (grosser) [inline lib/active_record/comments.rb]: no need to call it twice

File: spec/active_record/comments_spec.rb [hunk 3]
```diff
@@ -72,6 +72,26 @@ def capture_sql
       end
       sql.should == 'SELECT * FROM "users" WHERE "users"."id" = 1 /* xxx */'
     end
+
+    it "should be thread safe" do
+      res = []
+      ["xx", "yy"].map do |comment|
```
Comment (grosser) [inline spec/active_record/comments_spec.rb]: this test might also randomly be green, how about:

```Ruby
other = Thread.new { ActiveRecord::Comments.comment("xxx") { capture_sql { sleep 0.2; User.count } } }
sleep 0.1 # make other thread start
main = ActiveRecord::Comments.comment("yyy") { capture_sql { sleep 5; User.count } }
main.should_not include "xxx"
main.should include "yyy"
other.value.should_not include "yyy"
other.value.should_include "xxx"
```

**Comment by grosser** [comment]
:+1:

File: spec/active_record/comments_spec.rb (hunk 3 above)
Comment (osheroff) [inline spec/active_record/comments_spec.rb]: i checked, test definitely fails without the patch.  not sure why you say it might be randomly green?

Comment (grosser) [inline spec/active_record/comments_spec.rb]: ahhh I misread that the sleep makes sure both get started :)
(from my experience 0.1 is good enough)

File: spec/active_record/comments_spec.rb [hunk 4]
```diff
@@ -72,6 +72,26 @@ def capture_sql
...
+        Thread.new do
+          sql = nil
+          ActiveRecord::Comments.comment(comment) do
```
Comment (grosser) [inline spec/active_record/comments_spec.rb]: .comment is supposed to return what the yield returns, if it does not do that plz fix

**Comment by osheroff** [comment]
tests updated, can you merge on green?

**Comment by grosser** [comment]
:+1: :ship:

**Comment by osheroff** [comment]
slowest build ever is green.  (I can't merge or release)

**Comment by grosser** [comment]
public servers are hammered :/

**Comment by grosser** [comment]
0.1.3

you are contrib + gem owner now too :)

**Pull request merged**